blockchain/reactor_fsm.go

```diff
@@ -0,0 +1,430 @@
+package blockchain
+
+import (
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/tendermint/tendermint/libs/log"
+	"github.com/tendermint/tendermint/p2p"
+	"github.com/tendermint/tendermint/types"
+)
+
+// Blockchain Reactor State
+type bReactorFSMState struct {
+	name string
+
+	// called when transitioning out of current state
+	handle func(*bReactorFSM, bReactorEvent, bReactorEventData) (next *bReactorFSMState, err error)
+	// called when entering the state
+	enter func(fsm *bReactorFSM)
+
+	// timeout to ensure FSM is not stuck in a state forever
+	// the timer is owned and run by the fsm instance
+	timeout time.Duration
+}
+
+func (s *bReactorFSMState) String() string {
+	return s.name
+}
+
+// Blockchain Reactor State Machine
+type bReactorFSM struct {
+	logger log.Logger
+	mtx    sync.Mutex
+
+	startTime time.Time
+
+	state      *bReactorFSMState
```

Comment:
It is surprising that state of FSM is bReactorFSMState. I would expect state to hold protocol specific variables.